Hll Sketch and Theta sketch estimate can now be used as an expression

somu-imply opened the pull request:
HllSketchEstimate and Theta sketch estimates have been used only as PostAggs as they work on groupBy and TopN queries. But postAggs are not supported for scan queries. Window functions require these sketches estimates to work over a window engulfed by a scan query which requires these postAggs to be used as an expression. In this PR we introduce `HLL_SKETCH_ESTIMATE` and `THETA_SKETCH_ESTIMATE` as expressions. These estimates work on a sketch column and has the same behavior as the postAggs. New test cases have been added to show how scan queries can work with these estimates.

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

imply-cheddar commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllPostAggExprMacros.java:
```diff
@@ -0,0 +1,89 @@
+
+public class HllPostAggExprMacros
+{
+  // make same name as SQL binding
```
I think this comment was something you did for yourself?  Or is this intended for future readers of the code?

imply-cheddar commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllPostAggExprMacros.java:
```diff
@@ -0,0 +1,89 @@
+        public ExprEval eval(ObjectBinding bindings)
+        {
+          boolean round = false;
+          Expr arg = args.get(0);
```
No need to `.get(0)` on every single call here.  Do it once and reuse the reference.

imply-cheddar commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllPostAggExprMacros.java:
```diff
@@ -0,0 +1,89 @@
+          boolean round = false;
+          Expr arg = args.get(0);
+          ExprEval eval = arg.eval(bindings);
+          if (args.size() == 2) {
+            round = args.get(1).eval(bindings).asBoolean();
+          }
```
Also, no reason to check this on every single call.  `args` is gonna be the same for all of them.

clintropolis replied:
i think the issue is not having separate classes to use based on the number of arguments `apply` gets, splitting these up i think should resolve most of these sorts of issues

imply-cheddar commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllPostAggExprMacros.java:
```diff
@@ -0,0 +1,89 @@
+          if (valObj == null) {
+            return ExprEval.of(null);
+          }
+          if (eval.type().is(ExprType.COMPLEX) && valObj instanceof HllSketchHolder) {
```
Do you have a test that shows that `eval.type().is()` is necessary?  As long as it got an `HllSketchHolder`, I'm nto sure the type of teh `eval` really matters.

imply-cheddar commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllPostAggExprMacros.java:
```diff
@@ -0,0 +1,89 @@
+            return ExprEval.of(null);
+          }
+          if (eval.type().is(ExprType.COMPLEX) && valObj instanceof HllSketchHolder) {
+            HllSketchHolder h = HllSketchHolder.fromObj(valObj);
```
We already did an `instanceof` check for `HllSketchHolder` can't we just cast it?  Or, you could alternatively just pass in the `eval.value()` (assuming it's non-null) and just assume it's a type that this can deal with.

imply-cheddar commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllPostAggExprMacros.java:
```diff
@@ -0,0 +1,89 @@
+            double estimate = h.getEstimate();
+            return round ? ExprEval.of(Math.round(estimate)) : ExprEval.of(estimate);
+          } else {
+            throw HLLSketchEstimateExprMacro.this.validationFailed("requires a HllSketch as the argument");
```
`validationFailed` is not correct.  This is running in `eval()` which is actually operating on top of the data.  If you are trying to do a validation, that should in the `ExprMacro`'s `apply()` method.  Once we are here, we better be dealing with an object type that we can actually do something with.  I think the best implementation here is to just check if the `valObj` was null and if it is not, pass it into `HllSketchHolder.fromObj()` and just hope for the best.

imply-cheddar commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllPostAggExprMacros.java:
```diff
@@ -0,0 +1,89 @@
+    @Override
+    public Expr apply(List<Expr> args)
+    {
+      class HllSketchEstimateExpr extends ExprMacroTable.BaseScalarMacroFunctionExpr
```
What is the importance of doing this as an inner class like this?  I don't know that it adds anything and I think it just makes things more confusing.  I realize you were perhaps just copying a pattern that was already in place.  I looked and it looks like I copied there pattern here too 

https://github.com/apache/druid/pull/14195/files#diff-5c5c183e6c105fdbb9abfb3e63cb0e75e19b932c1767d1d4c8b7f6131aa3f7d5R52-R57

But, like, the lines highlighted there really show why this pattern doesn't make sense.  I'm calling `args.get(0)` even though I've already gotten the arg out and am additionally passing it to the constructor.  So there's just a ton of referencing the same thing through all sorts of different means, which is confusing.  Let's not copy this pattern around anymore.  Either make this a static class and have it store the arg that it needs.  Or just return it as an anonymous class.

imply-cheddar commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/sql/ThetaPostAggMacros.java:
```diff
@@ -0,0 +1,82 @@
+
+import java.util.List;
+
+public class ThetaPostAggMacros
```
I had various commentary in the HLL version.  Please make sure to apply it here as well.

imply-cheddar commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/sql/ThetaSketchEstimateOperatorConversion.java:
```diff
@@ -40,7 +40,7 @@
 public class ThetaSketchEstimateOperatorConversion implements SqlOperatorConversion
 {
-  private static final String FUNCTION_NAME = "THETA_SKETCH_ESTIMATE";
+  private static final String FUNCTION_NAME = "theta_sketch_estimate";
```
once again we have this casing change...  Is that important?

imply-cheddar commented on extensions-core/datasketches/src/test/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllSketchSqlAggregatorTest.java:
```diff
@@ -888,4 +902,187 @@ public void testGroupByAggregatorDefaultValuesFinalizeOuterSketches()
+                .build()
+        ),
+        ImmutableList.of(
+            new Object[]{0.0D},
+            new Object[]{1.0D},
+            new Object[]{1.0D},
+            new Object[]{1.0D},
+            new Object[]{1.0D},
+            new Object[]{1.0D}
```
I don't believe that any of the results actually changed once you turned on rounding...  It would be great to have this test run against some data set that actually makes a difference rounding and not rounding.

For that matter, once you do that, you can use the same expression twice in the same query, once that does the default, once that rounds and once that doesn't round and then validate the results (and show that they are different)

imply-cheddar commented on extensions-core/datasketches/src/test/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllSketchSqlAggregatorTest.java:
```diff
@@ -888,4 +902,187 @@ public void testGroupByAggregatorDefaultValuesFinalizeOuterSketches()
+          ImmutableList.of()
+      );
+    }
+    catch (ExpressionValidationException e) {
+      Assert.assertTrue(
+          e.getMessage().contains("Function[HLL_SKETCH_ESTIMATE] requires a HllSketch as the argument")
+      );
+    }
```
Once you adjust the expression as requested, I expect this test to change some.  If you do the "pass to HllSketchHolder and hope" route, I expect that you will get an IllegalArgumentException saying that the values of dim2 are not legal.  But it might be something else too.

imply-cheddar commented on extensions-core/datasketches/src/test/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllSketchSqlAggregatorTest.java:
```diff
@@ -888,4 +902,187 @@ public void testGroupByAggregatorDefaultValuesFinalizeOuterSketches()
+  public void testHllEstimateAsVirtualColumnWithGroupByOrderBy()
+  {
+    skipVectorize();
+    cannotVectorize();
```
Do you know what makes this not vectorizable?

clintropolis replied:
you have to [implement](https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/math/expr/Expr.java#L183) a [vector processor](https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/math/expr/vector/ExprVectorProcessor.java#L29) for expression to be vectorizable

imply-cheddar commented on extensions-core/datasketches/src/test/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllSketchSqlAggregatorTest.java:
```diff
@@ -888,4 +902,187 @@ public void testGroupByAggregatorDefaultValuesFinalizeOuterSketches()
+  public void testHllEstimateAsVirtualColumnWithTopN()
+  {
+    testQuery(
+        "SELECT"
+        + " HLL_SKETCH_ESTIMATE(hllsketch_dim1)"
+        + " FROM druid.foo"
+        + " GROUP BY 1 ORDER BY 1"
+        + " LIMIT 2",
```
Maybe throw a `COUNT(*)` on this and order by that instead?  Just to make it seem more "topN" like.

imply-cheddar commented on extensions-core/datasketches/src/test/java/org/apache/druid/query/aggregation/datasketches/theta/sql/ThetaSketchSqlAggregatorTest.java:
```diff
@@ -1018,8 +1034,160 @@ public void testGroupByAggregatorDefaultValuesFinalizeOuterSketches()
+                .build()
+        ),
+        ImmutableList.of(
+            NullHandling.replaceWithDefault() ? new Object[]{null} : new Object[]{0.0D},
```
What causes the difference here?  It would appear that this and HLL are not consistent with each other and I'm curious why.

clintropolis commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllSketchEstimateOperatorConversion.java:
```diff
@@ -43,7 +43,7 @@
 public class HllSketchEstimateOperatorConversion implements SqlOperatorConversion
 {
-  private static final String FUNCTION_NAME = "HLL_SKETCH_ESTIMATE";
+  private static final String FUNCTION_NAME = "hll_sketch_estimate";
```
it isn't case sensitive here, [we lowercase them all prior to adding to the map](https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/math/expr/ExprMacroTable.java#L56), however we do not lower-case them while looking them up when parsing the expression, https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/math/expr/ExprListenerImpl.java#L327 for macros, though this might technically be a bug because the built-in functions we do attempt to lowercase them first when looking up https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/math/expr/Parser.java#L90

clintropolis replied:
btw, this can just extend `DirectOperatorConversion` and lowercase the function name (or not if we modify it to not be case sensitive for macros, though we should be thoughtful before making such a change)

clintropolis commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllPostAggExprMacros.java:
```diff
@@ -0,0 +1,89 @@
+            throw HLLSketchEstimateExprMacro.this.validationFailed("requires a HllSketch as the argument");
+          }
+        }
+
```
ideally should implement `getOutputType` so this can behave better with all the type inference stuff

imply-cheddar commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/sql/ThetaPostAggMacros.java:
```diff
@@ -0,0 +1,92 @@
+    @Override
+    public Expr apply(List<Expr> args)
+    {
+      validationHelperCheckArgumentCount(args, 1);
+      return new ThetaSketchEstimateExpr(args.get(0));
```
Probably nice to add the "round" option while we are at it, makes it consistent with the HLL version.

somu-imply replied:
There are 2 options for this:
1. We only add the round to the expression and do not add it to the aggregator factory. Whatever result is returned by the aggregator factory is rounded (or not) and shown in the output.
2. We push the round as a parameter to the SketchMergeAggregatorFactory, compute it in the `finalizeComputation` method, and return the value as is.

1 is the easier way and does not affect many test cases, but 2, I believe is the better way. @imply-cheddar what do you suggest?

somu-imply replied:
After discussion we decided to leave PostAggs as is and make the expressions similar only

imply-cheddar replied:
Hrm, I'm okay with only implementing round in the expression, but in that case, because the expression and the PostAggregator are not consistent, you will need to stop planning the post aggregator from SQL.  I'm like 90% certain that the SQL planner is fine using an expression anywhere a post agg could do, it just "prefers" the PostAgg in certain circumstances.  So, it should be acceptable to change the SqlOperatorConversion to always return `null` from the call to build the PostAggregator and things should continue to just work.

Note, that the PostAggregator cannot be removed for compatibility reasons as it could be used from the native query APIs.

clintropolis replied:
i suppose could fall back to the expression post aggregator and use the rounding expression

imply-cheddar commented on extensions-core/datasketches/src/test/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllSketchSqlAggregatorTest.java:
```diff
@@ -888,4 +910,193 @@ public void testGroupByAggregatorDefaultValuesFinalizeOuterSketches()
+        "SELECT"
+        + " HLL_SKETCH_ESTIMATE(hllsketch_dim1), COUNT(*)"
+        + " FROM druid.foo"
+        + " GROUP BY 1 ORDER BY 1"
```
ORDER BY 2?

imply-cheddar commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/sql/ThetaPostAggMacros.java:
```diff
@@ -38,8 +39,9 @@ public static class ThetaSketchEstimateExprMacro implements ExprMacroTable.ExprM
-      validationHelperCheckArgumentCount(args, 1);
-      return new ThetaSketchEstimateExpr(args.get(0));
+      validationHelperCheckAnyOfArgumentCount(args, 1, 2);
+      //validationHelperCheckArgumentCount(args, 1);
```
commented out code?  Is this needed?

imply-cheddar commented on extensions-core/datasketches/src/test/java/org/apache/druid/query/aggregation/datasketches/theta/sql/ThetaSketchSqlAggregatorTest.java:
```diff
@@ -1073,6 +1081,38 @@ public void testThetaSketchEstimateAsVirtualColumn()
+  public void testThetaSketchEstimateAsVirtualColumn2Args()
+  {
+    testQuery(
+        "SELECT"
+        + " THETA_SKETCH_ESTIMATE(thetasketch_dim3, FALSE)"
+        + " FROM foo",
```
I don't expect this table to actually have enough items to generate numbers that would need rounding.  You should be able to access the wikipedia data set in the tests, which you should be able to get to estimate things.

Note that the smallest logk for Theta sketch allowed is 4, which is 16 entries.  So, you would have to do a theta sketch aggregation with a size of 16 and then see more than 16 distinct values in order to get it to estimate (and generate a true floating point number)

somu-imply replied:
Thanks, I am using a bigger dataset for the `true` case will update shortly

somu-imply commented:
After discussing with @cheddar I have reverted the change that made the syntax of HLL_SKETCH_ESTIMATE and THETA_SKETCH_ESTIMATE the same. Making that required removing a context parameter `sqlFinalizeOuterSketches` which was introduced for backward compatibility. Also keeping the expression and post agg different and making the post agg null for Theta Sketch Operator conversion, introduced issues with expressions over post aggs with theta_sketch_intersect and theta_sketch_estimate. Those changes are a lot to block this PR and those should be handled separately.

clintropolis commented on extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/hll/sql/HllPostAggExprMacros.java:
```diff
@@ -0,0 +1,100 @@
+    {
+      boolean round = false;
+      ExprEval eval = estimateExpr.eval(bindings);
+      if (isRound != null) {
```
this should be split into two classes, one for estimate and one for rounded estimate. otherwise we have to execute this if statement every row for something that is constant for the entire query

imply-cheddar replied:
Technically, if someone wanted the determination of rounding to be passed in by a different column, the code as defined here would be correct no?  I guess the question is whether we believe that the argument is a literal and therefore can be extracted once and reused or if we believe that it can be a reference.  

Do we have a good way of identifying that for expressions?

clintropolis replied:
yeah, there is `Expr.isLiteral` and `Expr.getLIteralValue` that can be used to check if a value is a constant

cheddar reviewed (approved):
There's an optimization left that could be done around identifying when the "round" parameter is a literal and then not checking it on every single evaluation.  The tests are passing though and the optimization is generally minor (this is expect to most often run in a post-agg context, so not that often), so I'm gonna go ahead and approve and merge as is for now.

Really though, the more I think about rounding, the more I think that SQL already has a round function, we should perhaps just follow a path of requiring people to wrap the output in a round function instead of adding parameters to expressions to do the rounding...
